src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs

```diff
@@ -561,6 +655,46 @@ private static bool TryParseValue(ref Utf8JsonReader reader, [NotNullWhen(true)]
             return new JsonDocument(utf8Json, database, extraRentedBytes);
         }
 
+        private static JsonDocument ParseUnrented(
+            ReadOnlyMemory<byte> utf8Json,
+            JsonReaderOptions readerOptions,
+            JsonTokenType tokenType)
+        {
+            // These tokens should already have been processed.
+            Debug.Assert(
+                tokenType != JsonTokenType.Null &&
+                tokenType != JsonTokenType.False &&
+                tokenType != JsonTokenType.True);
+
+            ReadOnlySpan<byte> utf8JsonSpan = utf8Json.Span;
+            MetadataDb database;
+
+            if (tokenType == JsonTokenType.String ||
+                tokenType == JsonTokenType.Number)
+            {
+                // For primitive types, we can avoid renting and there is no need for a StackRowStack.
+                database = MetadataDb.CreateLocked(utf8Json.Length);
+                var _ = new StackRowStack(initialSize: -1);
+                Parse(utf8JsonSpan, readerOptions, ref database, ref _);
+            }
+            else
+            {
+                database = MetadataDb.CreateRented(utf8Json.Length, convertToAlloc: true);
+                var stack = new StackRowStack(JsonDocumentOptions.DefaultMaxDepth * StackRow.Size);
+
+                try
+                {
+                    Parse(utf8JsonSpan, readerOptions, ref database, ref stack);
```

Comment:
How's test coverage for all these changes? Are we still at 100% for JsonElement/JsonDocument?

Reply:
The only new LOC that isn't hit is this branch: https://github.com/dotnet/runtime/pull/42538/files#diff-8bee6ed7017869dde2a37df1271e0338R612.
```cs
                if (valueSpan.IsEmpty)
                {
                    owned = valueSequence.ToArray();
                }
```

This branch is only hit if we expose this functionality publicly (not called from the serializer) since "valueSpan" will never be empty when using a reader instance with the serializer.

So we could remove this branch and add an Assert, but I suggest leaving until we decide on https://github.com/dotnet/runtime/issues/42755 which at that point we will add tests to cover that.